### Crear sus propios vectores con Gensim basado en lo visto en clase con otro dataset. 
### Probar términos de interés y explicar similitudes en el espacio de embeddings. 
### Intentar plantear y probar tests de analogías. 
### Graficar los embeddings resultantes.
### Sacar conclusiones.

In [6]:
import pandas as pd
from gensim.models import Word2Vec
# from keras.rc.legacy.preprocessing.text import text_to_word_sequence
from tensorflow.keras.preprocessing.text import text_to_word_sequence

2024-11-28 16:15:51.588846: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2024-11-28 16:15:51.885632: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2024-11-28 16:15:51.887770: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-11-28 16:15:53.087273: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


### Datos
Utilizaremos como dataset que contiene reseñas de usuarios de amazon a productos de software. (https://snap.stanford.edu/data/web-Amazon.html)

In [4]:
software_txt = pd.read_json('data/Software.json', lines=True)
software_txt = software_txt['reviewText']

In [5]:
software_txt[1]

'I am really enjoying this book with the worksheets that make you review your goals, what to do when you do not make it, it reminds me  of my human sexuality classwork.'

### 1 - Preprocesamiento

In [6]:
sequence_token = []

for txt in software_txt:
    try:
        sequence_token.append(text_to_word_sequence(txt))
    except Exception as e:
        continue

In [8]:
print(sequence_token[9])

['recieved', 'in', 'a', 'timely', 'manner', 'book', 'in', 'great', 'condition', 'some', 'markings', 'but', 'ledgible', 'will', 'be', 'ordering', 'again', 'from', 'this', 'site', 'for', 'my', 'school', 'books', 'saved', 'over', '200']


In [12]:
from gensim.models.callbacks import CallbackAny2Vec
# Durante el entrenamiento gensim por defecto no informa el "loss" en cada época
# Sobrecargamos el callback para poder tener esta información
class callback(CallbackAny2Vec):
    """
    Callback to print loss after each epoch
    """
    def __init__(self):
        self.epoch = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        if self.epoch == 0:
            print('Loss after epoch {}: {}'.format(self.epoch, loss))
        else:
            print('Loss after epoch {}: {}'.format(self.epoch, loss- self.loss_previous_step))
        self.epoch += 1
        self.loss_previous_step = loss

### 2 - Crear los vectores (word2vec)

In [9]:
# Crearmos el modelo generador de vectores
# En este caso utilizaremos la estructura modelo Skipgram
w2v_model = Word2Vec(min_count=5,    # frecuencia mínima de palabra para incluirla en el vocabulario
                     window=2,       # cant de palabras antes y desp de la predicha
                     vector_size=10,       # dimensionalidad de los vectores 
                     negative=20,    # cantidad de negative samples... 0 es no se usa
                     workers=4,      # si tienen más cores pueden cambiar este valor
                     sg=1)     

In [10]:
w2v_model.build_vocab(sequence_token)

In [15]:
# Cantidad de filas/docs encontradas en el corpus
print("Cantidad de docs en el corpus:", w2v_model.corpus_count)

Cantidad de docs en el corpus: 459370


In [11]:
# Cantidad de words encontradas en el corpus
print("Cantidad de words distintas en el corpus:", len(w2v_model.wv.index_to_key))

Cantidad de words distintas en el corpus: 39347


### 3 - Entrenar embeddings

In [17]:
# Entrenamos el modelo generador de vectores
# Utilizamos nuestro callback
w2v_model.train(sequence_token,
                 total_examples=w2v_model.corpus_count,
                 epochs=20,
                 compute_loss = True,
                 callbacks=[callback()]
                 )

Loss after epoch 0: 54491584.0
Loss after epoch 1: 21842856.0
Loss after epoch 2: 11545440.0
Loss after epoch 3: 11491640.0
Loss after epoch 4: 11414016.0
Loss after epoch 5: 11250904.0
Loss after epoch 6: 11143136.0
Loss after epoch 7: 1038152.0
Loss after epoch 8: 0.0
Loss after epoch 9: 0.0
Loss after epoch 10: 0.0
Loss after epoch 11: 0.0
Loss after epoch 12: 0.0
Loss after epoch 13: 0.0
Loss after epoch 14: 0.0
Loss after epoch 15: 0.0
Loss after epoch 16: 0.0
Loss after epoch 17: 0.0
Loss after epoch 18: 0.0
Loss after epoch 19: 0.0


(539796545, 720411360)

In [12]:
#w2v_model.save('model/software_model.pkl')

In [7]:
w2v_model = Word2Vec.load('model/software_model.pkl')

### 4 - Ensayar

In [28]:
# Palabras que MÁS se relacionan con...:
w2v_model.wv.most_similar(positive=['great'], topn=5)

[('fantastic', 0.9793065786361694),
 ('terrific', 0.9753286242485046),
 ('wonderfull', 0.9697865843772888),
 ('exellent', 0.9693095684051514),
 ('fabulous', 0.9657824039459229)]

Como se puede observar, buscamos las palabras más similares a 'great', obteniendo palabras tales como 'fantastic', 'terrific', 'wonderfull', etc, lo que significda que agrupa palabras que tienen significados positivos similares en las reviews. Este resultado nos indica que el modelo aprendió patrones linguistico de los corpus.

In [34]:
# Palabras que MÁS se relacionan con...:
w2v_model.wv.most_similar(positive=['green'], topn=5)

[('yellow', 0.9900084137916565),
 ('white', 0.9835501909255981),
 ('grey', 0.981337308883667),
 ('purple', 0.9773489832878113),
 ('greenish', 0.9720497727394104)]

De los resultados de obtener las palabras más similares a 'green' observamos que agrupas palabras que describen colores.

In [35]:
w2v_model.wv.similarity('software', 'bug')

0.6191907

Observando similitudes entre palabras

In [36]:
w2v_model.wv.similarity('price', 'discount')

0.91236776

El valor de similitud 0.912 indica que el modelo ha capturado bien las relaciones semánticas en las reviews. Esto tiene mucho sentido, ya que en el contexto de reviews de productos, es común que las palabras relacionadas con precios y descuentos aparezcan juntas en frases.

##### Pruebas de analogía

In [46]:
tests = [
    ("good", "excellent", "bad"),  
    ("product", "quality", "worth"),
    ("user", "password", 'license')
]

for a, b, c in tests:
    if all(word in w2v_model.wv for word in [a, b, c]):
        result = w2v_model.wv.most_similar(positive=[b, c], negative=[a])
        print(f"{a} -> {b} :: {c} -> {result[0][0]}")
    else:
        print(f"Uno o más términos ({a}, {b}, {c}) no están en el vocabulario.")

good -> excellent :: bad -> horrendous
product -> quality :: worth -> ratio
user -> password :: license -> authorization


- El modelo identifica que "excellent" es un superlativo de "good", una relación positiva. Para el término "bad", el modelo sugiere "horrendous", que intensifica negativamente el significado de "bad". Esto indica que el modelo está capturando la estructura de intensificación semántica. 
- En el segundo caso, el modelo ha devuelto una relación coherente, ya que "license" se asocia estrechamente con "authorization" de manera similar a cómo "user" se vincula con "password". Esta relación tiene sentido, ya que las palabras están relacionadas con conceptos de acceso y permisos en un sistema.

#### Podemos concluir que el modelo aprendió correctamente las conexiones entre estas palabras. 

### 5 - Visualizar agrupación de vectores

In [8]:
from sklearn.decomposition import IncrementalPCA
from sklearn.manifold import TSNE
import numpy as np

def reduce_dimensions(model, num_dimensions = 2 ):

    vectors = np.asarray(model.wv.vectors)
    labels = np.asarray(model.wv.index_to_key)

    tsne = TSNE(n_components=num_dimensions, random_state=0)
    vectors = tsne.fit_transform(vectors)

    return vectors, labels

In [18]:
# Graficar los embedddings en 2D
import plotly.graph_objects as go
import plotly.express as px

vecs, labels = reduce_dimensions(w2v_model)

MAX_WORDS=400
fig = px.scatter(x=vecs[:MAX_WORDS,0], y=vecs[:MAX_WORDS,1], text=labels[:MAX_WORDS])
fig.show() 

In [19]:
# Graficar los embedddings en 3D

vecs, labels = reduce_dimensions(w2v_model,3)

fig = px.scatter_3d(x=vecs[:MAX_WORDS,0], y=vecs[:MAX_WORDS,1], z=vecs[:MAX_WORDS,2],text=labels[:MAX_WORDS])
fig.update_traces(marker_size = 2)
fig.show() # esto para plotly en colab